examples/ssd/train_multi.py

```diff
@@ -0,0 +1,234 @@
+    # where n_positive is the total number of positives in the batch
+    def __next__(self):
+        batch = next(self._iterator)
+        n_positive = sum((mb_label > 0).sum() for _, _, mb_label in batch)
+        return [(img, mb_loc, mb_label, n_positive)
+                for img, mb_loc, mb_label in batch]
+
+    @property
+    def epoch(self):
+        return self._iterator.epoch
+
+    @property
+    def epoch_detail(self):
+        return self._iterator.epoch_detail
+
+    @property
+    def previous_epoch_detail(self):
+        return self._iterator.previous_epoch_detail
+
+    @property
+    def is_new_epoch(self):
+        return self._iterator.is_new_epoch
+
+    def finalize(self):
+        self._iterator.finalize()
+
+    def serialize(self, serializer):
+        self._iterator.serialize(serializer)
+
+    def reset(self):
+        self._iterator.reset()
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '--model', choices=('ssd300', 'ssd512'), default='ssd300')
+    parser.add_argument('--batchsize', type=int, default=32)
+    parser.add_argument('--test_batchsize', type=int, default=32)
+    parser.add_argument('--gpu', type=int, nargs='+')
+    parser.add_argument('--out', default='result')
+    parser.add_argument('--resume')
+    args = parser.parse_args()
+
+    if args.model == 'ssd300':
+        model = SSD300(
+            n_fg_class=len(voc_bbox_label_names),
+            pretrained_model='imagenet')
+    elif args.model == 'ssd512':
+        model = SSD512(
+            n_fg_class=len(voc_bbox_label_names),
+            pretrained_model='imagenet')
+
+    model.use_preset('evaluate')
+    train_chain = MultiboxTrainChain(model)
+    chainer.cuda.get_device_from_id(args.gpu[0]).use()
+
+    train = TransformDataset(
+        ConcatenatedDataset(
+            VOCBboxDataset(year='2007', split='trainval'),
+            VOCBboxDataset(year='2012', split='trainval')
+        ),
+        Transform(model.coder, model.insize, model.mean))
+    train_iter = TransformIterator(
+        chainer.iterators.MultiprocessIterator(train, args.batchsize))
+
+    test = VOCBboxDataset(
+        year='2007', split='test',
+        use_difficult=True, return_difficult=True)
+    test_iter = chainer.iterators.SerialIterator(
+        test, args.test_batchsize, repeat=False, shuffle=False)
+
+    # initial lr is set to 1e-3 by ExponentialShift
+    optimizer = chainer.optimizers.MomentumSGD()
+    optimizer.setup(train_chain)
+    for param in train_chain.params():
+        if param.name == 'b':
+            param.update_rule.add_hook(GradientScaling(2))
+        else:
+            param.update_rule.add_hook(WeightDecay(0.0005))
+
+    devices = {'main': args.gpu[0]}
+    devices.update(
+        ('other{:d}'.format(i + 1), gpu) for i, gpu in enumerate(args.gpu[1:]))
+    updater = training.ParallelUpdater(train_iter, optimizer, devices=devices)
```

Comment:
Do we need to make `train_multi.py` to behave the same way as `train.py`?
Does training get unstable if make changes?

Also, we should document some changes made from `train.py`, especially ones involving `n_positive`.

Reply:
> Do we need to make train_multi.py to behave the same way as train.py?
Does training get unstable if make changes?

I think it is bad that the number of GPUs affects the behaviour. We should keep the behaviour unchanged as much as possible. (I think it will converge even if we change the behaviour).
Do you want to change the behaviour according to the number of GPUs?

Reply:
> I think it is bad that the number of GPUs affects the behaviour.

Should we follow this rule for all multi-GPU implementations (e.g. ImageNet)?

> Do you want to change the behaviour according to the number of GPUs?

Well, I am wondering if we should use `ParallelUpdater`. Since this is the only option to implement the current behavior, we need to change the behavior if we use other updaters, such as ChainerMN and `MultiprocessParallelUpdater`.
In my opinion, we should always use the same updater for multi-GPU implementation, so that the community would not get confused about which one to use.

Is `ParallelUpdater` fast enough in this training script?
It can be slow because it does not use NCCL.

Reply:
> Should we follow this rule for all multi-GPU implementations (e.g. ImageNet)?

From my understanding, there is no concrete rule for multi-GPU implementations.
In addition, this implementation is not against to that of classification (softmax cross entropy does not change it's behaviour).

> Well, I am wondering if we should use ParallelUpdater. Since this is the only option to implement the current behavior, we need to change the behavior if we use other updaters, such as ChainerMN and MultiprocessParallelUpdater.

I agree that we should consider the cases of  ChainerMN or MultiprocessParallelUpdater.
I will check the APIs of them.

> Is ParallelUpdater fast enough in this training script?

It is not slow when it is compared with Caffe implementation. The biggest motivation of this implementation is to reduce the memory consumption per GPU (please refer Chainer-jp slack).

Reply:
> From my understanding, there is no concrete rule for multi-GPU implementations.

Yes. I would like to discuss if we are going to have any.

> In addition, this implementation is not against to that of classification (softmax cross entropy does not change it's behaviour).

Yeah, but BN needs to be modified if we want to make them the same.

Reply:
BN in ChainerMN keeps the behaviour, doesn't it?

Reply:
I think `MultiNodeBatchNormalization` is not common in ImageNet training protocols, such as the one in 1-hour paper.
So the behavior would change.

Reply:
I heard that PSPNet requires `MultiNodeBatchNormalization` for good performance. We cannot follow the protocol of ImageNet in that case. We don't need to treat `ImageNet in 1 hour` as a standard of all tasks/models.

Reply:
I think the best way is to follow the original papers as much as possible. Changing the behaviour is inconvenient for academic researches because we/users cannot estimate the amount of side-effect of modification.